### DISCUSSION TITLE
How to derive a property from a sum of a property within the items of ObservableCollection in a Reactive way?

### DISCUSSION THREAD

**Discussion by ProbablePrime** [AvaloniaUI/Avalonia#10978] [Q&A] [answered]
I'm working on a system that has a list of Records displayed in a list.

| Name    | Should be Included| Size |
|---------|-------------------|------|
| Record 1| [ ]               |5     |
| Record 2| [ ]               |5     |

Each item has the following properties:
- Name
- Id
- Size
- A Boolean indicating if this item should be included or not - ShouldBeIncluded

A user can check a checkbox to change the boolean of the item from false to true. etc.

Based on a user changing various item's checkboxes I need to keep a "TotalSize" property up to date so that it reflects the sum of all item's size property in the list where the item's "ShouldBeIncluded" property is true.

I'm trying to do this within an Avalonia based app running ReactiveUI/Dynamic Data in an MVVM way. 

For the purposes of this question I've minimised this all down to a simple reproduction though which can be found [here](https://github.com/ProbablePrime/Repro-ReactiveUI-MVVM-Sum-Filter/)

I think I need to Observe property changes on the ObservableCollection but I cannot figure out how to then bind this back to the ReactiveUI property.

- I've tried various DynamicData methods, including the ones I've commented out.
- I've tried ObservableAsAProperty helpers but I don't know how to construct the observable.
- I've tried doing this manually, but I want to keep in an MVVM/Reactive way.

But I want to understand the most Idiomatic way to do this, but everything I'm trying to do seems to overcomplicate it.

You can find the main heavy lifters in the only MVVM View Model [here](https://github.com/ProbablePrime/Repro-ReactiveUI-MVVM-Sum-Filter/blob/main/AvaloniaApplication6/ViewModels/MainWindowViewModel.cs)

![](https://i.probableprime.co.uk/230411-Mq5Zkl8eWq8AuLZFI2Bj.gif)

**Comment by alxgenov** [comment, marked as answer]
With [DynamicData](https://github.com/reactivemarbles/DynamicData) you can do tasks like this with absurdly minimal code like this:

```
private ObservableAsPropertyHelper<int> total;
public int Total => total.Value;

// In constructor
total = Records
    .ToObservableChangeSet(x => x.Id) 
    .AutoRefresh(x => x.ShouldBeIncluded)
    .Filter(x => x.ShouldBeIncluded)
    .Sum(x => x.Size)
    .ToProperty(this, x => x.Total);
```

**Reply by ProbablePrime** [reply]
Beautiful, This looks like what I was after, I'll take a look at integrating this and i'll get back to you with the results. You have made my day.

**Reply by ProbablePrime** [reply]
![image](https://user-images.githubusercontent.com/8791132/231300709-fa772871-8d14-4c50-b52a-237881dbab9e.png)

We were quite close it seems. looks like Sum is trying to sum the changeset and not the elements in the collection, I'll play around a bit more with the start of your example to see if I can get unblocked thank you!

**Reply by ProbablePrime** [reply]
I got it! Just needed a few more using statements. You can see them here: https://github.com/ProbablePrime/Repro-ReactiveUI-MVVM-Sum-Filter/commit/55d1042e52a5da6f773e0dc37f099226e27b9829

Marking as answer, thank you!
